### PR TITLE
Recommends lazily evaluated alternatives for `Option::or` and `Result::or`

### PR DESCRIPTION
**Pull request by jonathanstrong**
Adds language to docs for `Option` and `Result` recommending the use of lazily evaluated alternatives when appropriate. These comments are intended to echo a [clippy lint] on the same topic. The [reddit discussion] may also be of interest. 

[clippy lint]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#or_fun_call
[reddit discussion]: https://www.reddit.com/r/rust/comments/7hutqn/perils_of_optionor_and_resultor/

### CODE REVIEW

File: src/libcore/option.rs [hunk 1]
```diff
@@ -338,6 +338,10 @@ impl<T> Option<T> {
 
     /// Returns the contained value or a default.
     ///
+    /// Arguments passed to `unwrap_or` are eagerly evaluated; if you are passing 
+    /// the result of a function call, it is recommended to use `unwrap_or_else`,
```
Comment (kennytm) [inline src/libcore/option.rs]: Could you link to the `unwrap_or_else` method? i.e.

```
/// ... it is recommended to use [`unwrap_or_else`],
///
/// ...
///
/// [`unwrap_or_else`]: #method.unwrap_or_else
```

Similar for the other additions.

File: src/libcore/result.rs [hunk 2]
```diff
@@ -690,6 +694,10 @@ impl<T, E> Result<T, E> {
     /// Unwraps a result, yielding the content of an [`Ok`].
     /// Else, it returns `optb`.
     ///
+    /// Arguments passed to `unwrap_or` are eagerly evaluated; if you are passing
```
Comment (kennytm) [inline src/libcore/result.rs]: There is a trailing whitespace on this line, causing the CI to reject this PR. Other places affected:

```
[00:04:39] tidy error: /checkout/src/libcore/result.rs:697: trailing whitespace
[00:04:39] tidy error: /checkout/src/libcore/result.rs:698: trailing whitespace
[00:04:39] tidy error: /checkout/src/libcore/result.rs:699: trailing whitespace
[00:04:39] tidy error: /checkout/src/libcore/option.rs:341: trailing whitespace
[00:04:39] tidy error: /checkout/src/libcore/option.rs:342: trailing whitespace
[00:04:39] tidy error: /checkout/src/libcore/option.rs:343: trailing whitespace
[00:04:41] some tidy checks failed
```

File: src/libcore/option.rs (hunk 1 above)
Comment (jonathanstrong) [reply src/libcore/option.rs]: sure - will do

File: src/libcore/result.rs (hunk 2 above)
Comment (jonathanstrong) [reply src/libcore/result.rs]: fixed

**Review by dtolnay — APPROVED** [review]

**Pull request merged**